TEP-0156 : when-in-steps

ericzzzzzzz opened the pull request:
# Description

This pull request proposes the ability to define When Expressions directly within Steps. This allows for fine-grained control over step execution based on conditions like the status of previous steps, enabling more dynamic and adaptable workflows.

# Overview
 - Steps accept an optional when field with a list of When Expressions.
 - When Expressions are evaluated at the entrypoint level, allowing access to previous step results.
 - Skipped steps due to When Expressions have an exit code of 0 and a TerminationReason of Skipped. 
 - Both Operator-based and CEL expressions are supported.
 - Variable substitution within When Expressions includes access to Task Results, Params, Workspaces, and Step Results.

khrm reviewed (commented):
/kind tep

chitrangpatel commented:
/kind tep

chitrangpatel commented:
/assign

chitrangpatel commented on teps/0155-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,110 @@
+
+### Supported Expression Languages
+
+Just like when expressions in Task Level, when expressions in Step Level will also be supporting both Operator-based expressions and CEL(Common Expression Language) expressions. 
```
May be also add a paragraph on which fields are we applying when expressions on?
- all task based fields(e.g. Task Results, Task Conditions etc.)
- StepResults
- StepStatus(?): `TerminationReasonFields: Error, Continued, Skipped, TimeOut, Cancelled etc.`

ericzzzzzzz replied:
I added a `Variable Substitution` section to explains what which fields can be referenced in `step.when`.
 > StepStatus(?)
 It seems that we haven't StepStatus implemented yet, should we include in this tep? This seems a different scope.

chitrangpatel replied:
I'm ok keeping it out for now.

haroonc replied:
IMHO, we should include some way to allow conditional execution on step status. Otherwise usage of When expression in guarding a `step` will be limited. I think one major advantage of a When expression guarding a step is to check the status of preview steps. Otherwise we are forcing the user to "mimic" this functionality by exporting a pseudo status of step to workspace/result to mimic this functionality. 

Secondly, is it possible to check attributes of files in a workspace in When expressions?

ericzzzzzzz replied:
Hi, @haroonc 
Thanks for the review! I've added step status variable substitution when evaluating When expressions in the design. 
For checking file attributes, as far as i know Tekton doesn't support that directly. But, you can use a script in a step to get attributes and store them as results. Then, use those results in When expressions in later steps.

chitrangpatel commented:
@ericzzzzzzz can you please leave a short description of the TEP in the opening comment?

chitrangpatel commented:
@tektoncd/core-maintainers PTAL 🙏

ericzzzzzzz commented:
> @ericzzzzzzz can you please leave a short description of the TEP in the opening comment?

done! Thanks

afrittoli commented:
/assign

JeromeJu commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,116 @@
+          image: alpine
+          when:
+            - cel: "'monday'=='friday'"
+              script: |
```
Should this script be under step level?

ericzzzzzzz replied:
ahh ! yes! yes!

JeromeJu commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,116 @@
+          script: |
+             echo "wooooooo"
+          when:
+             - input: "$(steps.produce-step.results.result2)"
```
Are we not creating a DAG at `Step` level currently?

ericzzzzzzz replied:
steps are executed in-order, so they're naturally a DAG

chitrangpatel commented:
/assign @vdemeester 
PTAL since you had suggested opening this. 🙏

khrm commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,116 @@
+ - End-to-end tests covering common use cases and error conditions.
+
+## Alternatives
+An alternative approach would be to evaluate When Expressions before creating the pods for the steps. If a When Expression evaluates to false, the corresponding step container would not be created at all.
```
I think you meant `creating step containers inside the pods for the Task`.

khrm commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,116 @@
+
+Just like when expressions in Task Level, when expressions in Step Level will also be supporting both Operator-based expressions and CEL(Common Expression Language) expressions. 
+**Examples**
+ - Operator-based when expressions in steps.
```
Do we have any plans to deprecate this operator-based approach? Anything that can be done with an operator can also be done with `cel`.

chitrangpatel replied:
I don't know if we can do that since it is part of the `v1` spec already. API compatibility policy forbids it.

vdemeester commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,116 @@
+
+## Summary
+This TEP proposes the addition of When Expressions to individual Steps within a Task. This will allow for fine-grained control over the execution of steps based on conditions like the status of previous step results. This enhancement will provide greater flexibility and control in defining task logic, enabling more complex and dynamic workflows.
+## Motivation
```
Anything described here can be achieved by "relatively" smart scripts in Steps, am I right ?
The "main" difference is in one case it's expressed with the tekton API, and on the other case, it's in "a script" or "in code".

chitrangpatel replied:
Only for inlined steps. This allows us to control the execution of stepActions where we don't have the ability to edit the scripts.

vdemeester replied:
Yes, I can hear that 😛 But one could also "condition" this in the `StepAction` with a parameter (probably a bit convoluted)

haroonc replied:
That is a valid point, however, one can make same argument about `when` expression in Tasks 😉. 
IMHO, declarative conditionals in a CI/CD pipeline are more readable and improve maintainability of the pipeline.

vdemeester replied:
@haroonc on the other hand, it make the "split" between local build and CI/CD bigger 😛 but that's a completely different topic/debate 👼🏼

vdemeester reviewed (commented):
The lines between `Task` and `Pipeline` are bluring more and more with this. 
An additionnal question, what happens if a step that write a `TaskResult` is skipped ?

chitrangpatel commented:
> The lines between `Task` and `Pipeline` are bluring more and more with this. An additionnal question, what happens if a step that write a `TaskResult` is skipped ?

I think the core difference between the fact that Pipeline can fan-out (Matrix, DAG) vs Task is sequential will always remain. So it might appear that lines are being blurred but what you can do with Pipelines vs Tasks will always be different. If Tasks was a DAG of Steps then I agree that the lines would be truly blurred.


I think the outcome will be the same as if the Step producing a TaskResult errored out. The Task author is controlling the flow of the StepAction/Step that is producing the result so they can choose. In this case, I think that the Task will error with the message that it could not produce a result (whatever the current behavior is if the Task does not generate a result)?

vdemeester commented:
> I think the outcome will be the same as if the Step producing a TaskResult errored out. The Task author is controlling the flow of the StepAction/Step that is producing the result so they can choose. In this case, I think that the Task will error with the message that it could not produce a result (whatever the current behavior is if the Task does not generate a result)?

Can we discuss this in the TEP ?

ericzzzzzzz commented:
> > I think the outcome will be the same as if the Step producing a TaskResult errored out. The Task author is controlling the flow of the StepAction/Step that is producing the result so they can choose. In this case, I think that the Task will error with the message that it could not produce a result (whatever the current behavior is if the Task does not generate a result)?
> 
> Can we discuss this in the TEP ?

I added a section to capture this.

vdemeester reviewed (commented):
Sorry, one more thing, can we dig a bit more into examples of use case it would fullfill ?

For example
- having "optional" set of caching steps (using StepAction) if a parameter is the `Task`
- having a "report" and fail step if the previous step fails
- … other use case we envision with this feature.

ericzzzzzzz commented:
> Sorry, one more thing, can we dig a bit more into examples of use case it would fullfill ?
> 
> For example
> 
> * having "optional" set of caching steps (using StepAction) if a parameter is the `Task`
> * having a "report" and fail step if the previous step fails
> * … other use case we envision with this feature.

Hi, I added some examples.

vdemeester commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,235 @@
+
+### When Expressions Evaluation
+
+When expressions are evaluated at the entrypointer level. This allows us to use step results produced from previous steps to determine if a step should be skipped. To indicate a skipped step, the exit code will be 0 (showing a step is complete) and the TerminationReason will be Skipped.
```
Do we forsee any impact on the entrypoint "performance" ? The entrypoint is doing more and more work (with step results, …) and I think we never measured or "thought" about possible performance impact (and if they are noticeable or not) ?

ericzzzzzzz replied:
You're right, adding Step-level When Expressions will have some cost, but it's important to consider the value it brings. This cost is only incurred when Step-level When Expressions are actually configured. In comparison to the benefits of having this finer-grained control over workflow and the potential to skip long-running steps for performance optimization, the cost is relatively minimal. Even Task-level When Expressions also have a performance cost, and the Step-level implementation offers a significantly greater potential for performance gains through more flexible workflow control.

vdemeester commented on teps/0156-whenexpressions-in-step.md:
````diff
@@ -0,0 +1,235 @@
+        echo "reporting timeout!"
+```
+
+Here, the `report-timeout` step will only run if the `main-task` step terminates with a "TimeoutExceeded" status.  This allows you to specifically handle failures without executing the reporting logic every time.
````
In that case, the `Task` would still be marked as a failure, isn't it ?
If yes, is there a case where we wouldn't want that behavior ?

haroonc replied:
One potential use case could be a task that performs multiple integration test via separate stepActions for each integration service. We may want to file a bug report for each integration failure, mainly because those will be investigate by different teams. Report filing might be another step Action (e.g. jira-open-bug). We can declaratively control bug filing for each integration run via `when` expression.

ericzzzzzzz replied:
I just changed this example a little bit, it turns out with the existing implementation, timeout is a error that will always fail subsequent steps, https://github.com/tektoncd/pipeline/blob/9c315909a373a600eb42f343584badeea818607c/pkg/entrypoint/entrypointer.go#L255-L257 . 

Task authors should use "onError:continue" on the steps that could potentially fail to allow the follow-up steps to execute, authors can decide if the task should fail on the step handling the error.

ericzzzzzzz commented:
Hi @vdemeester , I think I've addressed all the comments you provided on tep. Could you please take another look and let me know if there's anything else you'd like to discuss?
Once you're happy with the tep, could you please provide your approval? Thanks!

vdemeester commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,116 @@
+ - Task-based fields:
+   - Task Results: You can access the results of any previously executed Task within the same TaskRun using the syntax $(tasks.<taskName>.results.<resultName>).
+ - Workspaces: You can reference the names of workspaces declared in the Task using the syntax $(workspaces.<workspaceName>.bound). This allows you to check if a workspace is bound and make decisions based on its availability.
+ - Params: Access input parameters defined in the Task using the familiar syntax $(params.<paramName>). This enables conditional logic based on the values passed into the Task.
```
I think in theory yes.

chitrangpatel commented:
Thanks for your work @ericzzzzzzz and thanks for the detailed review Vincent!!

/lgtm

afrittoli commented on teps/0156-whenexpressions-in-step.md:
```diff
@@ -0,0 +1,237 @@
+
+### Impact On Task Results
+When a step designed to produce a TaskResult is skipped due to a failing when expression, it should be treated like a task that produces results being skipped in a pipeline.
+The Task author is responsible for controlling the flow of the steps, including the production of Task Results, and the when expression acts as a mechanism to control this flow. Subsequent steps relying on an unavailable result  due to step being skipped might encounter errors. The Task author should carefully consider the dependencies between steps and ensure the Task logic can handle scenarios  ...
```
TEP-0048 would be a nice help for this :)
